[BitcoinCash branch] cashaddr

ptschip opened the pull request.

luke-jr committed:
Qt: Add "Copy URI" to payment request context menu

deadalnix committed:
Do not delete config.h on clean

Summary: When building in tree, make clean deletes src/config.h .

Test Plan:
Make a build in tree, run

  make clean

Checks that config.h is still there.

Reviewers: freetrader, #bitcoin_abc

Reviewed By: freetrader, #bitcoin_abc

Differential Revision: https://reviews.bitcoinabc.org/D39

schancel committed:
bitcoin URI to bitcoincash URI

Summary: Defined const URI_SCHEME in GUIUtil namespace, and updated associated usage of the bitcoin scheme to use this URI_SCHEME constant.

Test Plan:
* Manually checking the GUI and other portions of the client for bitcoin -> bitcoincash change.
* Run GUI Tests

Reviewers: deadalnix, zquestz, O1 Bitcoin ABC, #bitcoin_abc, CCulianu, freetrader

Reviewed By: deadalnix, zquestz, O1 Bitcoin ABC, #bitcoin_abc, CCulianu, freetrader

Subscribers: dgenr8, zander, freetrader, CCulianu

Tags: #bitcoin_abc

Maniphest Tasks: T88

Differential Revision: https://reviews.bitcoinabc.org/D466

sipa committed:
Introduce wrappers around CBitcoinAddress

Summary:
This patch removes the need for the intermediary Base58 type
CBitcoinAddress, by providing {Encode,Decode,IsValid}Destination
function that directly operate on the conversion between strings
and CTxDestination.

This is a port of https://github.com/bitcoin/bitcoin/pull/11117

Depends on D539

Test Plan: make check, rpc tests, manual QT testing

Reviewers: #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Subscribers: deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D540

dooglus committed:
qt: Use IsMine to validate custom change address

Summary: Port of https://github.com/bitcoin/bitcoin/pull/11247

Test Plan: As part of next diff

Reviewers: #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Subscribers: deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D539

deadalnix committed:
Add an utility to convert string of bits from base 2^N to 2^M .

Summary:
This is extracted from D544 .

based on code from Pieter Wuille.

Test Plan: Added unit tests.

Reviewers: dagurval, #bitcoin_abc, tomtomtom7

Reviewed By: dagurval, #bitcoin_abc, tomtomtom7

Differential Revision: https://reviews.bitcoinabc.org/D572

deadalnix committed:
Cashaddr decoding the raw data and checksum

Summary:
This is split from D544 plus implement a candidate BCH code of degree 8 for it. The BCH code can detect up to 6 errors over the typical length of an address and up to 8 errors in a row.

This is intended to serve as a base to get the new address format started.

Test Plan: Added unittests.

Reviewers: #bitcoin_abc, dagurval

Reviewed By: #bitcoin_abc, dagurval

Subscribers: dagurval

Differential Revision: https://reviews.bitcoinabc.org/D576

dagurval committed:
Add cashaddr prefix to chainparams

Summary: This is extarcted from D544 .

Test Plan: \o/

Reviewers: #bitcoin_abc, dagurval

Reviewed By: #bitcoin_abc, dagurval

Differential Revision: https://reviews.bitcoinabc.org/D610

deadalnix committed:
Use standard prefix in cashaddr tests

Summary: As per title.

Test Plan:
  make check

Reviewers: dagurval, schancel, #bitcoin_abc

Reviewed By: schancel, #bitcoin_abc

Differential Revision: https://reviews.bitcoinabc.org/D620

dagurval committed:
Move uint160S to header

Summary: Make these functions reusable.

Test Plan: None

Reviewers: #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D631

dagurval committed:
Support for encoding addresses using cashaddr.

Summary: Depends on D631

Test Plan: Added unittests.

Reviewers: #bitcoin_abc, schancel, deadalnix

Reviewed By: #bitcoin_abc, schancel, deadalnix

Subscribers: deadalnix, schancel

Differential Revision: https://reviews.bitcoinabc.org/D623

deadalnix committed:
Split the decoding of cashaddr in a pure decode step + a Destination creation step

Summary: This allows for much more in depth testing, even for destination we don't have support for.

Test Plan: Added a bunch of unittests.

Reviewers: dagurval, schancel, freetrader, #bitcoin_abc

Reviewed By: dagurval, #bitcoin_abc

Differential Revision: https://reviews.bitcoinabc.org/D652

dagurval committed:
Enable decoding and optional encoding of cashaddr

Summary:
Wraps both cashaddr and base58 addresses into EncodeDestination and
DecodeDestination.

Test Plan: Added unittest

Reviewers: #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Subscribers: deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D660

sipa committed:
format for base58

dagurval committed:
Use bch prefix in cashaddr

Summary: Reflects changes in https://github.com/Bitcoin-UAHF/spec/pull/31

Test Plan: make check

Reviewers: #bitcoin_abc, Mengerian, schancel

Reviewed By: #bitcoin_abc, Mengerian, schancel

Differential Revision: https://reviews.bitcoinabc.org/D682

dagurval committed:
Added DummyConfig

Summary: Implementation of Config with dummy values. Subclassing this simplifies creating config objects in unittests.

Test Plan: make check

Reviewers: #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Subscribers: deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D685

dagurval committed:
Add flag to enable dst encoding using cashaddr

Summary:
Adds option -usecashaddr to enable encoding using cashaddr. Defaults to
false.

RPC tests always use the new cashaddr format.

Test Plan: make check, rpc tests

Reviewers: #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D695

dagurval committed:
Make DummyAddress support cashaddr

Summary: Depends on D685

Test Plan: Added unittest.

Reviewers: #bitcoin_abc, freetrader

Reviewed By: #bitcoin_abc, freetrader

Differential Revision: https://reviews.bitcoinabc.org/D553

dagurval committed:
Support cashaddr in parse/format URI

Summary: Depends on D708

Test Plan: added tests

Reviewers: #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Subscribers: deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D709

dagurval committed:
Make Config more available in Qt

Test Plan: ran qt client

Reviewers: #bitcoin_abc, deadalnix, schancel

Reviewed By: #bitcoin_abc, deadalnix, schancel

Subscribers: deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D708

dagurval committed:
Support cashaddr in Qt input validation

Summary: This makes Qt address input fields accept cash addresses.

Test Plan: added unittest, manual test

Reviewers: #bitcoin_abc, deadalnix, freetrader

Reviewed By: #bitcoin_abc, deadalnix, freetrader

Subscribers: freetrader, deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D552

deadalnix committed:
Remove redundant code in cashaddr tests

Summary: As per title. It's doing the same computation twice.

Test Plan:
  make check

Reviewers: dagurval, schancel, #bitcoin_abc

Reviewed By: schancel, #bitcoin_abc

Differential Revision: https://reviews.bitcoinabc.org/D746

schancel committed:
Encode cashaddr size properly in version bit.

Summary: Encode the hash size in the version bit

Test Plan: make -j8 src/test/test_bitcoin && src/test/test_bitcoin --run_test='cashaddrenc_tests/*'

Reviewers: deadalnix, freetrader, dagurval, #bitcoin_abc

Reviewed By: deadalnix, #bitcoin_abc

Differential Revision: https://reviews.bitcoinabc.org/D761

schancel committed:
Fix cashaddrenc_tests/check_size

Summary: Fix the construction logic for check_size test.

Test Plan: make check

Reviewers: deadalnix, dagurval, freetrader, #bitcoin_abc

Reviewed By: deadalnix, #bitcoin_abc

Differential Revision: https://reviews.bitcoinabc.org/D762

dagurval committed:
Encode destination using Base58 in WalletDB

Summary:
Previously destinations were encoded with EncodeDestination, which
depends on user configuration. WalletDB needs consistent encoding.

Base58 encoding is used for backward compatibility.

Tests for Purpose, Name and DestData in CWalletDB

Test Plan: added tests

Reviewers: #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Subscribers: deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D745

dagurval committed:
Accept multiple schemes in parseBitcoinURI

Test Plan: qt unittests

Reviewers: #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D769

deadalnix committed:
Fix walletdb_tests on windows

Summary:
Windows is a bit more restrictive when it comes to files not properly closed and alike. This caused an error in the cleanup after running hte test (the DB files were kept open and couldn't be deleted).

This ensure that the DB is closed properly after use. The use of MockDB ensure that files shared accross DB instance such as db.log do get mamaged in memory instead of on disk.

Test Plan:
Cross compile windows binaries, then

  wine src/test/test_bitcoin.exe -t walletdb_tests

Reviewers: schancel, #bitcoin_abc

Reviewed By: schancel, #bitcoin_abc

Differential Revision: https://reviews.bitcoinabc.org/D791

dagurval committed:
Modify QR code to fit cashaddr

Summary:
This bumps the size of the QR code and also decreases the font size when
using cashaddr.

The encoded address is displayed as text in the QR code image. This allows
for the longer cashaddr to fit in the image.

Test Plan: Create a request using QT client on mainnet. Tested on Ubuntu.

Reviewers: #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D802

dagurval committed:
Introduce bitcoinUriScheme

Summary: Adds support for the new uri schemes introduced by cashaddr.

Test Plan: Added tests

Reviewers: deadalnix, schancel, kyuupichan, #bitcoin_abc

Reviewed By: deadalnix, #bitcoin_abc

Differential Revision: https://reviews.bitcoinabc.org/D797

dagurval committed:
Support parsing cashaddr from command line

Summary:
This makes the client support parsing cashaddr encoded addresses for all networks.

Depends on D797

Test Plan:
  ./bitcoin-qt "bitcoincash:mz8DUu5atAfNEBNeQHrpdGkfWZ9vmk3Q13?amount=42.00000000&label=test1&message=a%20message"

Ensure that the client boot in testnet mode and the proper sent is prefilled.

  ./bitcoin-qt "bitcoincash:1F6MqpCw3f8JRErcZMYjkFLBzpFvvjobZZ?amount=42.00000000&label=mainnet1&message=a%20message"

Ensure the client boots in mainnet and the sent form is prefilled.

  ./bitcoin-qt "bchtest:qp5dzwdct8xcpec62z799rdfpylxftsl6v0trfz2m9?amount=42.00000000&label=cashtest2&message=a%20message" "bitcoincash:mz8DUu5atAfNEBNeQHrpdGkfWZ9vmk3Q13?amount=42.00000000&label=test1&message=a%20message"

Ensure that the client start on testnet and fill up the sent form properly.

  ./bitcoin-qt "bitcoincash:qrc5xvnqwv3esj8hs5w85gh3xjwgnnjfvv3ll6wmdu?amount=42.00000000&label=test3&message=a%20message" "bitcoincash:1F6MqpCw3f8JRErcZMYjkFLBzpFvvjobZZ?amount=42.00000000&label=mainnet1&message=a%20message"

dito

  ./bitcoin-qt "bchreg:qrxvv8fpv42wvlls6z4mp9qvrs3kar2pau5tssl8xd?amount=42.00000000&label=cashtest1&message=a%20message"

Start in regtest mode.

Reviewers: #bitcoin_abc, dagurval, schancel

Reviewed By: #bitcoin_abc, schancel

Subscribers: schancel, deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D583

dagurval committed:
Use user configured encoding in receive request

Summary:
The address associated with a receive request is encoded using the
configuration of the client at the time. This can be either base58 or
cashaddr.

When displaying an existing receive request, display with the currently
configured encoding.

Test Plan:
- Added test.
- Create two requests, one running with -usecashaddr and one without.
- Start client with and without -usercashaddr, observe that requests are displayed correct.

Reviewers: #bitcoin_abc, dagurval

Reviewed By: #bitcoin_abc, dagurval

Subscribers: deadalnix

Differential Revision: https://reviews.bitcoinabc.org/D801

deadalnix committed:
Allow cashaddr's prefix to be implicit

Summary: This makes for a much more pleasant UX.

Test Plan:
Added unittests.

In addition, try to send coin on testnet using both cashaddr with and without prefix and legacy addresses, with and without the `-usecashaddr` flag.

Reviewers: schancel, dagurval, #bitcoin_abc, Mengerian

Reviewed By: schancel, #bitcoin_abc, Mengerian

Subscribers: Mengerian

Differential Revision: https://reviews.bitcoinabc.org/D839

deadalnix committed:
Activate cashaddr on Jan, 14

Summary: This activate cashaddr by default on Jan, 14. The flag `-usecashaddr` can still be used to select manually which format is desired.

Test Plan: Run bitcoin-qt and check usecashaddr is disabled by default.

Reviewers: dagurval, schancel, Mengerian, #bitcoin_abc

Reviewed By: schancel, #bitcoin_abc

Differential Revision: https://reviews.bitcoinabc.org/D840

ptschip committed:
remove accidental text in chainparams.h

ptschip committed:
Fix compiler issues

BITCOIN_CONFIG_H is already being used and also fixed
a linker error.

ptschip committed:
Remove unnecessary declarations in config.h

ptschip committed:
Fix formatting

ptschip commented:
@gandrewstone @sickpig cashaddr is ready for testing

ptschip commented:
bitcoin-QT is crashing on startup, bitcoind is fine though.  Seems to be happening after
665e38e was applied.

ptschip committed:
fix QT crash on startup

also fix crash when clicking on request payment button

ptschip commented:
Fixed the crash on starup...testing around a bit and finding the coin freeze doesn't seem to be working when making a coin freeze address...i'll have to look into that next week.

ptschip committed:
Fix for freeze address not showing up in receiverequest dialog

gandrewstone commented on depends/README.md:
```diff
@@ -14,7 +14,7 @@ For example:
 A prefix will be generated that's suitable for plugging into Bitcoin's
 configure. In the above example, a dir named x86_64-w64-mingw32 will be
-created. To use it for Bitcoin:
+created. To use it for bitcoin-abc:
```
remove this change

ptschip replied:
done

gandrewstone commented on doc/release-notes.md:
```diff
@@ -5,6 +5,7 @@ Bitcoin Unlimited version 0.12.1 is now available from:
 This is a new minor version release, including ........,
 various bugfixes and updated translations.
 
+<<<<<<< HEAD
```
looks like merge annotation accidentally got in

ptschip replied:
fixed

gandrewstone commented on doc/release-notes.md:
```diff
@@ -122,3 +123,10 @@ Thanks to everyone who directly contributed to this release:
+ - Only connect to node using the cash magic.
+ - Remove indicator mentionning if a node uses the cash magic getpeerinfo RPC.
+ - Add support for the new cashaddr format. The `-usecashaddr` flag can be used to select which format is used when presenting addresses to users. By default, Bitcoin ABC will keep using the old format until Jan, 14 and then switch to the new format. Both format are now accepted as input.
+>>>>>>> 7557637... Activate cashaddr on Jan, 14
```
merge annotation and change Bitcoin ABC to "this client".

ptschip replied:
fixed

gandrewstone commented on src/bitcoin-tx.cpp:
```diff
@@ -298,19 +308,30 @@ static void MutateTxAddOutData(CMutableTransaction &tx, const string &strInput)
+
+    // Extract FLAGS
+    bool bScriptHash = false;
+    if (vStrInputParts.size() == 3)
```
If the string has 4 parts this code will ignore the flags.  To maintain compatibility with the previous behavior, we would ignore additional sections.  so this should be: if (vStrInputParts.size() > 2)

ptschip replied:
fixed

gandrewstone commented on src/config.h:
```diff
@@ -0,0 +1,45 @@
+    bool UseCashAddrEncoding() const override { return false; }
+};
+
+// Temporary woraround.
```
why is this a workaround, what should really be done?

ptschip replied:
I didn't put the text there so I don't know why it was considered a workaround and I don't see how it could be different...

dagurval replied:
Not my comment either, but I believe the intention is not have a global getter, but rather pass the global config instance as argument to everywhere it's needed.

gandrewstone replied:
I'm ok with using a global since I can't conceive of a good reason for running multiple configs simultaneously.  Let's remove the comment.

ptschip committed:
Remove reference to bitcoin-abc in /depends/README.md

ptschip committed:
Fix release-notes.md

ptschip committed:
Retain original logic in MutateTxAddOutScript

rather than check for vStrInputParts == 3 we want to use
> 2 in the event that the string ever has 4 parts.

sickpig commented on contrib/rpm/bitcoin.spec:
```diff
@@ -0,0 +1,440 @@
+%define bdbv 4.8.30
```
rpm spec file is not a bad add at all but I think we should mv it on its own PR

ptschip replied:
fixed

sickpig commented on doc/release-notes.md:
```diff
@@ -36,79 +37,27 @@ earlier.
 
 ### Block and transaction handling
 
+Remove the newdaaactivationtime configuration.
```
This change is not related to cashaddr

ptschip replied:
fixed

sickpig commented on doc/release-notes.md:
```diff
@@ -36,79 +37,27 @@ earlier.
 ### P2P protocol and network code
 
-The p2p alert system has been removed in #7692 and the 'alert' message is no longer supported.
+Only connect to node using the cash magic.
```
ditto

ptschip replied:
fixed.

sickpig commented on src/Makefile.am:
```diff
@@ -95,13 +97,15 @@ BITCOIN_CORE_H = \
   compat/endian.h \
   compat/sanity.h \
   compressor.h \
+  config.h \
```
why we need this one?

ptschip replied:
this is the new config that we had to carry over from ABC to make cashaddr work.

sickpig commented on src/Makefile.am:
```diff
@@ -477,7 +485,6 @@ clean-local:
 	-$(MAKE) -C secp256k1 clean
 	-$(MAKE) -C univalue clean
 	-rm -f leveldb/*/*.gcda leveldb/*/*.gcno leveldb/helpers/memenv/*.gcda leveldb/helpers/memenv/*.gcno
-	-rm -f config.h
```
seems wrong to me

ptschip replied:
config.h was being deleted when you ran "make clean"..this was incorrect and should not be happening.

sickpig commented on src/allowed_args.cpp:
```diff
@@ -439,7 +439,9 @@ static void addWalletOptions(AllowedArgs &allowedArgs)
-                _("(1 = keep tx meta data e.g. account owner and payment request information, 2 = drop tx meta data)"));
+                _("(1 = keep tx meta data e.g. account owner and payment request information, 2 = drop tx meta data)"))
+        .addArg("usecashaddr", optionalBool,
+            _("Use Bitcoin Cash Address for destination encoding (Activates by default Jan 14, 2017)"));
```
this means we need to release in a few days right?

ptschip replied:
yes, ABC is activating by default on the 14th.  I think we need to get this out there as well ASAP.

gandrewstone commented on src/cashaddr.cpp:
```diff
@@ -0,0 +1,304 @@
+ */
+std::pair<std::string, data> Decode(const std::string &str,
+                                    const std::string &default_prefix) {
+    // Go over the string and do some sanity checks.
```
what's the longest cashaddr?  Let's add this at the top:
if (str.size() > X) return {};

That way the rest of the routine is protected from someone sending a very long string...

ptschip replied:
I tried just using a length of 42 + prefixlength + 1 for the separator...it generally works but many of tests then fail.  Will have to research this a bit.

gandrewstone commented on src/qt/locale/bitcoin_be_BY.ts:
```diff
@@ -410,8 +414,8 @@
-        <source>Open a %1: URI or payment request</source>
-        <translation>Адкрыць %1: URI ці запыт плацяжу</translation>
+        <source>Open a %1 URI or payment request</source>
+        <translation>Адкрыць біткойн: URI ці запыт плацяжу</translation>
```
the %1 got lost...

ptschip replied:
fixed

gandrewstone commented on src/qt/locale/bitcoin_ko_KR.ts:
```diff
@@ -402,8 +402,8 @@
-        <source>Request payments (generates QR codes and %1: URIs)</source>
-        <translation>지불 요청하기 (QR코드와 비트코인이 %1: URIs)</translation>
+        <source>Request payments (generates QR codes and %1 URIs)</source>
+        <translation>지불 요청하기 (QR코드와 비트코인이 생성됩니다: URIs)</translation>
```
%1

ptschip replied:
fixed

gandrewstone commented on src/script/ismine.h:
```diff
@@ -19,7 +19,8 @@ class CScript;
     ISMINE_NO = 0,
-    //! Indicates that we don't know how to create a scriptSig that would solve this if we were given the appropriate private keys
+
+
```
probably good to keep this comment...

ptschip replied:
fixed

ptschip committed:
Replace appropriate coment that was accidentally deleted

ptschip committed:
Fix two translations - add missing %1

sickpig commented on src/chainparams.cpp:
```diff
@@ -265,6 +266,7 @@ class CUnlParams : public CChainParams
             boost::assign::list_of(0x42)(0x69)(0x67)(0x20).convert_to_container<std::vector<unsigned char> >();
```
why there's a different way to set up `EXT_PUBLIC_KEY` and `EXT_SECRET_KEY` in respect to what we do for bch testnet a few lines below (L368)?

ptschip replied:
It's an older definition..the ones carried over from ABC were newer and look better , but there's nothing wrong with what is there.  Maybe in a cleanup PR we can update the old definitions so everything is uniform.

sickpig commented on src/config.cpp:
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2017 The Bitcoin developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
```
wonder if we could put the content of this file in `src/globals.cpp` in a future PR?

ptschip replied:
Good question...I think it's intent is correct and helps to pass the correct config based on what environment we're running in rather than a global var which we need regardless of environment

sickpig commented on src/qt/guiconstants.h:
```diff
@@ -56,7 +56,7 @@ static const int TOOLTIP_WRAP_THRESHOLD = 80;
 
 /* QRCodeDialog -- size of exported QR Code image */
-#define EXPORT_IMAGE_SIZE 256
+#define QR_IMAGE_SIZE 350
```
why this change in image size?

ptschip replied:
it was a carry over from the ABC PR...i think it's fine and actually a bigger image is i think better.  I tried scanning the image from my computer with my cell and it works fine whereas in the past I sometimes had issues (if my memory serves)

dagurval replied:
In ABC, the text representation of the address is embedded in the QR code. The image needed to be increased to fit the text.

ptschip committed:
Add new files to formatted-files list

ptschip committed:
Remove release notes not related to cashaddr

ptschip committed:
Format cashaddrenc.cpp/h

ptschip committed:
Remove rpm/bitcoin.spec

This is not part of the cashaddr PR and was carried over
accidentally from the ABC client.

ptschip commented:
@sickpig ok, took out the bitcoin.spec.

gandrewstone commented:
@ptschip, in the send dialog, the Public Label field has an & in front of it.  This is supposed to indicate a hotkey but somehow it is being displayed instead.  We can fix this in a subsequent PR if nothing else pops up...
